### ISSUE TITLE
Add conditional tests in goss files

### ISSUE THREAD

**Issue by aya** [aelsabbahy/goss#136]
The short story : I would like to be able to load variables in my goss files.

The context :
Goss is played on local hosts, but combined with ansible it could be used to validate a set of servers, with 
Goss files sent and played on each remote servers.

The problem :
Tests sets should be adapted to specific operating systems.
If I want to validate that mysql client package is installed, I would check for 'mysql' package on redhat servers or 'mysql-client' package on debian servers.
If I want to validate a specific version, I would check for mysql-5.1.73 on centos 6 or mysql-client-5.6 on ubuntu 16.04.
In this example, at least the package name and the version should be readable from a variable if I want to keep one goss file for all my servers.
I may whish to activate tests under specific conditions, too. A variable {use_systemd} may lead to add more tests, or modify some tests for a specific server.

My solution :
Actualy, I'm able to [launch goss with ansible](https://github.com/aya/ansible-template/blob/master/tests/goss.yml) with a specific goss file for each OS.

My needs :
It would be awesome if goss was able to parse the var files of my ansible role.
So I would have the same input for both my role and the tests validating my role.
A variable {{ssh_enable}} set to true would allow ansible to start the sshd service and goss to test that sshd is listening on port {{ssh_port}}.
If set to false, it would be possible to skip the associated test with goss.

**Comment by aelsabbahy** [comment]
In the meantime I would love to get @taurus-forever and @mika feedback on this feature.

Basically, at a high level my thoughts are, a user is able to provide a yaml/json with variables, and the goss.yml files are treated as golang [text/template](https://golang.org/pkg/text/template/). This way the JSON/YAML can also be generated from specialized sys fact tools like factor or ohai... But I have a few concerns:

1. How useful is this? On one hand using text templating language to generate valid YAML/JSON can sometimes be painful, but this will handle both the conditional and variable use case. And other projects do this (ex. [kubernetes/helm](https://github.com/kubernetes/helm))

2. How have you found using external templating languages, I'm not sure how mature the golang templating language is since I haven't used it, but I know for example Ansible/Jinja2 is pretty flexible, will incorporating this in Goss be an improvement for Ansible users? There's also [http://jsonnet.org/](http://jsonnet.org/) which is very powerful, but comes with a steep learning curve.

Would love to get more feedback on where users feel the responsibility of `goss` begins and ends with regards to variable and conditional tests.

**Comment by aelsabbahy** [comment]
Ok, so sort of back from vacation, some thoughts:

So the template solution I'm thinking of would be something like this:

goss --vars somevars.{json,yaml} validate

Where somevars.json would be an arbitrary set of user provided variables (ex, output of [facter](https://github.com/puppetlabs/facter)):
```json
{
    "os": {
        "distro": {
            "id": "Fedora"
        }
    }
}
```

And the goss.yaml would be a golang text template, ex:
```yaml
service:
  {{- if eq .os.distro.id "Fedora"}}
    httpd:
  {{- else}}
    apache2:
  {{- end}}
    enabled: true
    running: true
```

This would allow for complex templating while keeping `goss` out of the business of generating vars from the server. A user would be able to use whatever external var file or tool they like by leveraging bash process substitution. For example, using puppetlabs facter would be:
```
goss --vars <(facter -j) validate
```

Thoughts?

**Comment by taurus-forever** [comment]
I would say the last option should do the job.
While you are opening Pandora's box here.

Adding non-linear logic will require better understanding what is going on inside,
kind of some debug/verbose mode. Otherwise it will be hard to understand what is going on inside the test:
```
> Total Duration: 0.123s
> Count: 195, Failed: 0, Skipped: 0
```
The output above doesn't allow you to trace which testcase has been executed, etc.

P.S. it might be great to have some ability to use env variables, like:

TEST_DNS=false goss validate

**Comment by lowks** [comment]
yes last options looks good

**Comment by CMeza99** [comment]
I want to run the same test to verify a service is reachable, acls/iptables are correct, etc in different subnets. Being able to pass such information would be very helpful.

**Comment by aya** [comment]
To complete your example of a templating solution, it would be nice to get the content of the variables available in the goss file :
```
service:
 {{- if eq .os.distro.id "Fedora"}}
  httpd:
 {{- else if eq .os.distro.id "Fedora"}}
  apache2:
 {{- end}}
    enabled: {{ .thisUserWillActivateItLater }}
    running: {{ .thisUserWantsApacheRunning }}
```

This would allow complex templating as well as simple goss files like :
```
service:
    {{ .os.distro.service }}:
      enabled: {{ .thisUserWillActivateItLater }}
      running: {{ .thisUserWantsApacheRunning }}
```

With an OS specific somevars.json :
```
{
    "os": {
        "distro": {
            "id": "Fedora"
            "service": "httpd"
        }
    }
    "thisUserWillActivateItLater" : true
    "thisUserWantsApacheRunning" : true
}
```

The need is to test both system (depends on remote OS) and user (depends on user configuration) specific rules. The ability to call goss with --vars option multiple time may be usefull to help running this kind of test.

**Comment by aelsabbahy** [comment]
Released in [v0.3.0](https://github.com/aelsabbahy/goss/releases/tag/v0.3.0)

**Comment by avishnyakov** [comment]
Hey guys, sorry for bringing this up.

Conditions in goss testing is definitely a good topic. Easy usage would be a great feature, so just wonder if adding something like puppet does with "onlyif / unless" makes sense?

In that case, early mentioned YAML would be even simpler. Say, skipping this under light docker container, or under RHEL6, executing only under Vagrant or RHEL7. Just a suggestion to see if this might be a good idea.

```yaml
service:
   my-service:
    enabled: true
    running: true
    onlyif: 'which systemctl && ! ls /.dockerenv'
```

facter usage works well but complicates some of the YAML templates, they might easily become unnecessary complex.
